### PR TITLE
models/user.py: Should check email form is valid.

### PR DESCRIPTION
**Pull request by sh92**

### CODE REVIEW

File: hyde/models/user.py [hunk 1]
```diff
@@ -1,4 +1,9 @@
 from ..extensions import db
+import re
+
+
+def isEmail(email):
```
Comment (corona10) [inline hyde/models/user.py]: This function should go to utils.py

File: tests/models/test_user.py [hunk 2]
```diff
@@ -20,10 +20,10 @@ def tearDown(self):
```
Comment (corona10) [inline tests/models/test_user.py]: Add a unit test function for which check invalidation of emails.

File: hyde/models/user.py [hunk 3]
```diff
@@ -12,11 +17,15 @@ class User(db.Model):
...
-        # TODO: Deny non-email form.
+        if not isEmail(git_email):
+            print('Not valid git email form')
+            return
         self.git_email = git_email
         self.git_name = git_name
         self.password = password
-        # TODO: Deny non-email form.
+        if not isEmail(user_email):
+            print('Not valid user email form')
```
Comment (corona10) [inline hyde/models/user.py]: Do not use print function.
Use exception.

File: hyde/models/user.py (hunk 1 above)
Comment (corona10) [inline hyde/models/user.py]: Also need a unit test for this function.

File: tests/models/test_user.py (hunk 2 above)
Comment (sh92) [inline tests/models/test_user.py]: Gotch. It's going on!

File: hyde/models/user.py (hunk 1 above)
Comment (sh92) [inline hyde/models/user.py]: Got it! So we need to test utils.py independently. Right?

File: tests/models/test_user.py (hunk 2 above)
Comment (corona10) [inline tests/models/test_user.py]: @sh92 Absolutely.

Comment (corona10) [inline tests/models/test_user.py]: In future, we need to change `raise Exception('Not valid git email form')` into `raise InvalidEmailException` something like that 👍

File: hyde/models/user.py (hunk 1 above)
Comment (sh92) [inline hyde/models/user.py]: Ok, I think it it useless hard work if I test all utils file. But, Making module is great idea!

File: tests/models/test_user.py (hunk 2 above)
Comment (sh92) [inline tests/models/test_user.py]: Thx!

File: tests/models/test_user.py [hunk 4]
```diff
@@ -20,10 +20,16 @@ def tearDown(self):
...
-        users.append(User('git_email2', 'git_name2',
-                          'password2', 'user_email2', 'user_name2'))
+        try:
+            users.append(User('git@email.com', 'name',
+                              'password', 'user@email.com', 'user_name'))
+            users.append(User('git@email2.com', 'name2',
+                              'password2', 'user2@email2.com', 'user_name2'))
+            users.append(User('gitemail3.com', 'name2',
+                              'password2', 'user2@email3com', 'user_name2'))
+        except Exception as e:
```
Comment (corona10) [inline tests/models/test_user.py]: This test is wrong.
We need to add separate test for validation test.
Validation failed will never occur.

Comment (corona10) [inline tests/models/test_user.py]: We can not proceed _eq_ test by this exception.
That will be a problem.

Comment (sh92) [inline tests/models/test_user.py]: How can I do that? if I test fail, I could not merge PR

**Comment by corona10** [comment]
@sh92 Make 2 unit test for this PR.
* InvalidAddUserTest
* TestInvalidEmailTest

File: hyde/utils/validate.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+import re
+
+def isEmail(email):
```
Comment (corona10) [inline hyde/utils/validate.py]: I prefer function name start with Upper case.
Like `IsEmail`.

Comment (corona10) [inline hyde/utils/validate.py]: But never mind it's up to you!

Comment (sh92) [inline hyde/utils/validate.py]: Why don't we use like that? variable -> low_low, mehtod -> lowUperr, Class -> UpperUperr

Comment (corona10) [inline hyde/utils/validate.py]: cool!

**Comment by sh92** [comment]
@corona10 PTAL

File: hyde/models/user.py [hunk 6]
```diff
@@ -12,11 +13,13 @@ class User(db.Model):
     user_name = db.Column(db.String(255), nullable=False)
 
     def __init__(self, git_email, git_name, password, user_email, user_name):
-        # TODO: Deny non-email form.
+        if not isEmail(git_email):
+            raise InvalidEmailException()
```
Comment (corona10) [inline hyde/models/user.py]: @sh92 Did you check `InvalidEmailException` is raised?

File: tests/models/test_user.py [hunk 7]
```diff
@@ -18,12 +18,38 @@ def setUp(self):
...
+                     'user2@email3.com', 'user_name2')
+            self.session.add(u)
+        except Exception as e:
+            pass
+
+        try:
+            u = User('git@email3.com', 'name2', 'password2',
+                     'user2email3.com', 'user_name2')
+            self.session.add(u)
+        except Exception as e:
```
Comment (corona10) [inline tests/models/test_user.py]: `Exception` should be `InvalidEmailException`.

Comment (sh92) [inline tests/models/test_user.py]: OK

**Comment by corona10** [comment]
@sh92 Last one!. 
` hyde/exception/exc.py` -> ` hyde/exception/exec.py` or ` hyde/exception/exception.py`

**Comment by corona10** [comment]
@sh92 LGTM

**Comment by corona10** [comment]
LGTM

**Pull request merged**